src/ui/public/saved_objects/saved_objects_client.ts

```diff
@@ -0,0 +1,358 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { cloneDeep, pick, throttle } from 'lodash';
+import { resolve as resolveUrl } from 'url';
+
+import {
+  MigrationVersion,
+  SavedObject as PlainSavedObject,
+  SavedObjectAttributes,
+  SavedObjectReference,
+  SavedObjectsClient as SavedObjectsApi,
+} from '../../../../src/server/saved_objects';
+import {
+  CreateResponse,
+  FindOptions,
+  UpdateResponse,
+} from '../../../../src/server/saved_objects/service';
+import { isAutoCreateIndexError, showAutoCreateIndexErrorPage } from '../error_auto_create_index';
+import { kfetch, KFetchQuery } from '../kfetch';
+import { keysToCamelCaseShallow, keysToSnakeCaseShallow } from '../utils/case_conversion';
+import { SavedObject } from './saved_object';
+
+interface RequestParams {
+  method: 'POST' | 'GET' | 'PUT' | 'DELETE';
+  path: string;
+  query?: KFetchQuery;
+  body?: object;
+}
+
+interface CreateOptions {
+  id?: string;
+  overwrite?: boolean;
+  migrationVersion?: MigrationVersion;
+  references?: SavedObjectReference[];
+}
+
+interface BulkCreateOptions<T extends SavedObjectAttributes = any> extends CreateOptions {
+  type: string;
+  attributes: T;
+}
+
+interface UpdateOptions {
+  version?: string;
+  migrationVersion?: MigrationVersion;
+  references?: SavedObjectReference[];
+}
+
+interface BatchResponse<T extends SavedObjectAttributes = any> {
```

Comment:
Would it make sense to default to `SavedObjectAttributes` here? Or is there a case where not every saved object would at least satisfy that interface?